src/libasr/pass/replace_symbolic.cpp

```diff
@@ -0,0 +1,360 @@
+#include <libasr/asr.h>
+#include <libasr/containers.h>
+#include <libasr/exception.h>
+#include <libasr/asr_utils.h>
+#include <libasr/asr_verify.h>
+#include <libasr/pass/replace_symbolic.h>
+#include <libasr/pass/pass_utils.h>
+#include <libasr/pass/intrinsic_function_registry.h>
+
+#include <vector>
+
+
+namespace LCompilers {
+
+using ASR::down_cast;
+using ASR::is_a;
+
+
+class ReplaceSymbolicVisitor : public PassUtils::PassVisitor<ReplaceSymbolicVisitor>
+{
+public:
+    ReplaceSymbolicVisitor(Allocator &al_) :
+    PassVisitor(al_, nullptr) {
+        pass_result.reserve(al, 1);
+    }
+    std::vector<std::string> symbolic_dependcies;
+
+    void visit_Function(const ASR::Function_t &x) {
+        // FIXME: this is a hack, we need to pass in a non-const `x`,
+        // which requires to generate a TransformVisitor.
+        ASR::Function_t &xx = const_cast<ASR::Function_t&>(x);
+        SymbolTable* current_scope_copy = this->current_scope;
+        this->current_scope = xx.m_symtab;
+        for (auto &item : x.m_symtab->get_scope()) {
+            if (ASR::is_a<ASR::Variable_t>(*item.second)) {
+                ASR::Variable_t *s = ASR::down_cast<ASR::Variable_t>(item.second);
+                this->visit_Variable(*s);
+            }
+        }
+        transform_stmts(xx.m_body, xx.n_body);
+
+        SetChar function_dependencies;
+        function_dependencies.n = 0;
+        function_dependencies.reserve(al, 1);
+        for( size_t i = 0; i < xx.n_dependencies; i++ ) {
+            function_dependencies.push_back(al, xx.m_dependencies[i]);
+        }
+        for( size_t i = 0; i < symbolic_dependcies.size(); i++ ) {
+            function_dependencies.push_back(al, s2c(al, symbolic_dependcies[i]));
```

Comment:
Remove all the strings from `symbolic_dependencies` after this or at the start of this `visit_Function`.
As you added `symbol_set` to the symtab, the `PassVisitor` visits that function as well.